modules/core/src/main/java/org/apache/ignite/internal/client/thin/ReliableChannel.java

```diff
@@ -681,6 +682,8 @@ synchronized void initChannelHolders() {
         if (idx != -1)
             currDfltHolder = holders.get(idx);
 
+        Set<ClientChannelHolder> uniqueHolders = new HashSet<>();
```

Comment:
Looks like, you don't need one more collection. You can base on existing logic:
1. There is a loop over existing holders (see line 653)
2. And there is a place where a new holder is created (see line 700)